### PR TITLE
[MPS] Add native `cumsum` implementation

### PR DESCRIPTION
**Pull request by malfet**
Using https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/4057333-cumulativesumwithtensor?language=objc

Fall back to CPU if running on older MacOS versions
In `unary_op` add output tensor dims/dtype to the graph key (as even in default op we check output graph type)
Also, upcast int16 to int32 as MPS cumsum op on Ventura returns incorrect results for Int16 type (and it makes total sense for int8, as chances for overflow are very high)

### CODE REVIEW

**Review by kulinseth — APPROVED** [review]

File: aten/src/ATen/native/mps/operations/UnaryOps.mm [hunk 1]
```diff
@@ -7,6 +7,13 @@
 #include <ATen/native/mps/OperationUtils.h>
 #include <torch/library.h>
 
+// TODO: Remove me when moved to MacOS 13
+@interface MPSGraph (VenturaOps)
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/UnaryOps.mm:11]: I think we should probably move this to a header file and include all the venturaOps in the same place.

File: aten/src/ATen/native/mps/operations/UnaryOps.mm [hunk 2]
```diff
@@ -263,5 +270,33 @@ void unary_op(const Tensor& self, const Tensor& output, std::string op_name, Una
                 });
 }
 
+
+static bool mpsSupportsCumsum() {
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/UnaryOps.mm:274]: This also probably belongs to a single place.

Comment (malfet) [reply aten/src/ATen/native/mps/operations/UnaryOps.mm:274]: Yes, I believe it should be property of MPSDevice and run only once. Do you mind if I'll do it as followup PR?

**Comment by qqaatw** [comment]
@malfet Hi, do we currently have a CI flow that runs on macOS Ventura?

**Comment by blackrabbit** [comment]
Thanks; this is awesome.

**Comment by KayKozaronek** [comment]
Is it correct to assume that 'aten::cumsum.out' has now been implemented? If so, which version of PyTorch am I supposed to run? I just updated PyTorch to version 1.13.1 and I'm still getting an error. 

> UserWarning: The operator 'aten::cumsum.out' is not currently supported on the MPS backend and will fall back to run on the CPU.

Should I run the Nightly version? 

I hope this is the right place to ask this.

**Comment by malfet** [comment]
@KayKozaronek as one can easily see https://github.com/pytorch/pytorch/commit/657f2e12f0e212b2f4afd89ab2c824c409dcc951 is not part of https://github.com/pytorch/pytorch/tree/release/1.13 so yes, it's only available in nightlies at the moment

**Comment by mayank31398** [comment]
I think cumsum supports dim=-1 but this PR doesn't for MPS device.
Can we fix? @malfet 
```shell
Traceback (most recent call last):
  File "/Users/mayankmishra/miniconda3/envs/ai/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/mayankmishra/miniconda3/envs/ai/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/mayankmishra/Desktop/IBM/transformers-bloom-inference/inference_server/cli.py", line 43, in <module>
    main()
  File "/Users/mayankmishra/Desktop/IBM/transformers-bloom-inference/inference_server/cli.py", line 36, in main
    response = model.generate(text=[input_text], generate_kwargs=generate_kwargs)
  File "/Users/mayankmishra/Desktop/IBM/transformers-bloom-inference/inference_server/model_handler/deployment.py", line 199, in generate
    raise response
  File "/Users/mayankmishra/Desktop/IBM/transformers-bloom-inference/inference_server/models/model.py", line 49, in generate
    output = self.model.generate(
  File "/Users/mayankmishra/miniconda3/envs/ai/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/Users/mayankmishra/miniconda3/envs/ai/lib/python3.9/site-packages/transformers/generation/utils.py", line 1518, in generate
    return self.greedy_search(
  File "/Users/mayankmishra/miniconda3/envs/ai/lib/python3.9/site-packages/transformers/generation/utils.py", line 2282, in greedy_search
    model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
  File "/Users/mayankmishra/miniconda3/envs/ai/lib/python3.9/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 999, in prepare_inputs_for_generation
    position_ids = attention_mask.long().cumsum(-1) - 1
RuntimeError: Expected dim to be between 0 and 2 but got -1
```

**Comment by malfet** [comment]
@mayank31398 thank you for reporting, but you should've created issue, so that more people would see it. Fixing in #94119

**Comment by nick-maykr** [comment]
Hey, @malfet, thanks for the fix. Sorry for the dumb question, but has it been merged already? I see that the bot attached a "Merged" label to it, but when I install pytorch-nightly the error still persists.
My torch.__version__  is '2.0.0.dev20230107'

**Comment by malfet** [comment]
@nick-maykr Not sure what fix you are referring to. Cumsum implementation or negative index? Former was landed back in 2022, but later just a few days back, so it couldn't have been part of Jan 3rd nightly build.

**Comment by nick-maykr** [comment]
@malfet I meant the second one, thanks! When I installed the nightly build this problem no longer occurs.

However, I'm facing a new one: `RuntimeError: MPS does not support min/max ops with int64 input`. This one seems to be on HF Transformers side and afaik, is has not been addressed yet :(

**Comment by kulinseth** [comment]
@nick-maykr , we are looking into working around this by casting it to int32 (with a warning raised), till we have support for it in MPS kernels to handle it properly. cc @albanD

**Pull request closed without merging**